jmespath-core/src/main/java/io/burt/jmespath/Adapter.java

```diff
@@ -82,6 +94,12 @@
   T getProperty(T value, String name);
 
   /**
+   * Returns all of the property names of the given object, or an empty list
+   * when the given value is not an object.
+   */
+  Collection<String> getPropertyNames(T value);
```

Comment:
My concern here is that with the current interface, a JRuby implementation would require converting from `Map<RubyString,RubyBasicObject>` to `Collection<String>` to `List<RubyString>` which on the one hand will create a temporary string that isn't useful, but also will have two `RubyString` objects for each key, as the connection between the original `RubyString` and the new returned in the array is lost. (Obviously the JRuby implementation would use `RubyHash` and `RubyArray` rather than `Map` and `List`, but I'm using the interfaces to be able to explicit about the values of keys and values.)

If the interface was instead `T getPropertyNames(T value)`, then the JRuby implementation would only require calling `keys()`, and the implementation of the `keys` function could just return the value as is. Even for Jackson, which actually uses strings for keys (and doesn't seem to have a built-in function for getting a `JsonNode` containing the keys), the adaptor implementation would be about the same length as the current implementation (without the extra overhead):

```java
ArrayNode names = JsonNodeFactory.instance.arrayNode();
if (isObject(value)) {
  for (String fieldName : value.fieldNames()) {
   names.add(fieldName);
 }
}
return names;
```